src/sdl_backend.c

```diff
@@ -346,35 +364,36 @@ void sdl_push_samples(struct sdl_backend* sdl_backend, const void* src, size_t s
     if (sdl_backend->error != 0)
         return;
 
-    /* XXX: it looks like that using directly the pointer returned by cbuff_head leads to audio "cracks"
-     * with better resamplers whereas adding cbuff.head inside each memcpy doesn't... Really strange !
-     */
-    cbuff_head(&sdl_backend->primary_buffer, &available);
-    unsigned char* dst = (unsigned char*)sdl_backend->primary_buffer.data;
+    /* truncate to full samples */
+    if (size & 0x3) {
+        DebugMessage(M64MSG_WARNING, "sdl_push_samples: pushing non full samples: %zu bytes !", size);
+    }
+    size = (size / 4) * 4;
 
+    /* We need to lock audio before accessing cbuff */
+    SDL_LockAudio();
+    unsigned char* dst = cbuff_head(&sdl_backend->primary_buffer, &available);
     if (size <= available)
     {
-        SDL_LockAudio();
-
         if (sdl_backend->swap_channels) {
```

Comment:
shouldn't this be:
if (!sdl_backend->swap_channels)

Reply:
I didn't touch that logic here, but you're right the logic is only good for LittleEndian host, for BigEndian host the logic should be !swap_channels.
I can do a separate PR for that.
Explanation:
- N64 produces 16bit L/R samples in RDRAM that we can name "Lh Ll Rh Rl" (Left high/low, Right high/low).
- mupen organize RDRAM such that a 32bit word can be read by the host in a single instruction (no swapping). So for BigEndian host, bytes are ordered the same and for LittleEndian host bytes are "Rl Rh Ll Lh".
- memcpy doesn't reorder the bytes, so for BigEndian host we end up with "Lh Ll Rh Rl" which is non-swapped, but "Rl Rh Ll Lh" for LittleEndian host which is swapped
- The other branch swaps consecutive 16-bit hwords, so for BigEndian host L/R get swapped, and for LittleEndian host we end up with "Ll Lh Rl Rh" which is non swapped L/R samples.
So yeah, logic is wrong for BigEndian host, but OK for LittleEndian host.

Reply:
See PR #37 for a possible fix